distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Period.enso

```diff
@@ -0,0 +1,146 @@
+
+       Arguments:
+       - other_period: The period to add to `self`. Note that this cannot be a
+         `Duration`, neither `Date_Time`.
+
+       > Example
+       Add 1 day to 1 month.
+
+           import Standard.Base.Data.Time.Period
+
+           example_add = (Period.new months=1) + (Period.new days=1)
+    + : Period -> Period ! (Time_Error | Illegal_Argument_Error)
+    + self other_period =
+        other_period_checked = ensure_period other_period
+        Panic.catch Any (Period.Period_Data (self.internal_period.plus other_period_checked.internal_period)) err->
+            case err of
+                DateTimeException -> Error.throw Time_Error_Data "Period addition failed"
+                ArithmeticException -> Error.throw Illegal_Argument_Error_Data "Arithmetic error"
+
+    ## Subtract a specified amount of time from this period.
+
+       Arguments:
+       - other_period: Other Period to add to this Period. Note that this
+         cannot be a `Duration`, neither `Date_Time`.
+
+       > Example
+         Subtract 11 months from a period of 3 years
+
+             import Standard.Base.Data.Time.Period
+
+             example_subtract = (Period.new years=3) - (Period.new months=11)
+
+    - : Period -> Period ! (Time_Error | Illegal_Argument_Error)
+    - self other_period =
+        other_period_checked = ensure_period other_period
+        Panic.catch Any (Period.Period_Data (self.internal_period.minus other_period_checked.internal_period)) err->
+            case err of
+                DateTimeException -> Error.throw Time_Error_Data "Period subtraction failed"
+                ArithmeticException -> Error.throw Illegal_Argument_Error_Data "Arithmetic error"
+
+    ## Check two periods for equality.
+
+       Note that two periods are equal if they have the exact same amount of
+       years, months, and days. So `(Period.new days=30)` and
+       `(Period.new months=1`) are not equal.
+
+       Arguments:
+       - other_period: The period to compare against `self`.
+    == : Period -> Boolean ! (Time_Error | Illegal_Argument_Error)
+    == self that =
+        other_period = ensure_period that
+        self.internal_period.equals other_period.internal_period
+
+    ## Compare this period to another period to produce an ordering.
+    compare_to : Period -> Ordering ! (Time_Error | Illegal_Argument_Error)
+    compare_to self that =
+        other_period = ensure_period that
+        if self.years > other_period.years then Ordering.Greater else
+            if self.years < other_period.years then Ordering.Less else
+                if self.months > other_period.months then Ordering.Greater else
+                    if self.months < other_period.months then Ordering.Less else
+                        if self.days > other_period.days then Ordering.Greater else
+                            if self.days < other_period.days then Ordering.Less else
+                                Ordering.Equal
```

Comment:
It looks like comparing two `Period`s is weird to say the least. Moreover, the comparison that I have implemented seems very wrong. I think that comparing Periods is reasonable use case. I suggest keeping also number of days for every `Period`. Currently, we keep only `java.time.Period` as a field. That way, we can always compare two Periods - just compare the total number of days. Moreover, after that, we could also have a method like `Period.get_total_days`, which could be exported. What do you think?

Reply:
I would be very much in favour of comparing `Period.get_total_days` (`Period.total_days`?) instead of those conditionals but I would keep in private. This is because it impossible to get a valid value for `months` and `years` without any context. But if we use the same multipliers (e.g. `month=30 days` and `year=365`) for internal calculations then we are OKish.

Reply:
As @hubertp mentioned `Period.total_days` depends on context - so I don't think exporting such a function makes sense.

As for comparisons using such a private approximation - I'm still skeptical, what do you guys @hubertp @Akirathan @jdunkerley think about my examples above? IMO Periods without context 'which date is being shifted' are just incomparable - if I have periods `A` and `B`, then for `x+A` and `x+B`, depending on the date `x`, the total shift may be greater for A, greater for B or equal for both. So IMO it just doesn't make sense to compare periods.

One last reason to compare them could be for usage in data structures like Map - but I don't imagine it's often required to have a Map whose _keys_ are periods. Still, ideally one should just be able to use a HashMap (once we have that) and not require to impose a fake-and-forced ordering on the Periods.

Reply:
@radeusgd There are two types of periods:
1) Created via `Period.between start_date end_date`, for which we can compute `total_days` or rather save `start_date`.
2) Created via `Period.new years months days`, for which we don't have any `start_date`.

IMHO, it is definitely a sane use-case to compare two Periods of the first type, e.g., if I have two employees with start and end date of their employments, I want to be able to check which of the employees was employed longer. Comparing the first type with the second type also makes sense, one use-case is, e.g., checking if an employee was employed for more than 3 years. Comparing two periods of the second type **is not sane**, that is an error.

My suggestion is:
- For Periods created via `Period.between`, save `start_date` as a field.
- The `Period.compare_to` works as follows:
  - If only one of the periods have `start_date`, then the result is equivalent with `(start_date + period1).compare_to (start_date + period2)`. Probably attach warning?
  - If both periods have `start_date`, then compute `end_date` for both periods, and compute `Duration` between end dates and start dates. The result is comparison of these two durations.
  - If neither have `start_date`, throw Error.